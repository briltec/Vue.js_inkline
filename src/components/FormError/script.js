import AttributesProviderMixin from '@inkline/inkline/mixins/components/providers/AttributesProviderMixin';
import ClassesProviderMixin from '@inkline/inkline/mixins/components/providers/ClassesProviderMixin';
import InjectParentFormProviderMixin from '@inkline/inkline/mixins/forms/providers/InjectParentFormProviderMixin';

import EmitInputMethodMixin from '@inkline/inkline/mixins/components/methods/EmitInputMethodMixin';

import DisabledPropertyMixin from '@inkline/inkline/mixins/forms/properties/DisabledPropertyMixin';
import LoadingPropertyMixin from '@inkline/inkline/mixins/components/properties/LoadingPropertyMixin';
import NamePropertyMixin from '@inkline/inkline/mixins/forms/properties/NamePropertyMixin';
import ParentFormGroupPropertyMixin from '@inkline/inkline/mixins/forms/properties/ParentFormGroupPropertyMixin';
import ReadonlyPropertyMixin from '@inkline/inkline/mixins/forms/properties/ReadonlyPropertyMixin';
import SizePropertyMixin from '@inkline/inkline/mixins/components/properties/SizePropertyMixin';
import TabIndexPropertyMixin from '@inkline/inkline/mixins/components/properties/TabIndexPropertyMixin';

export default {
    name: 'IFormGroup',
    props: {
        inline: {
            type: Boolean,
            default: false
        },
        validate: {
            type: Boolean,
            default: true
        }
    },
    data() {
        return {
            inputSchema: null,
            errors: []
        };
    },
    mixins: [
        AttributesProviderMixin,
        ClassesProviderMixin,
        InjectParentFormProviderMixin,

        EmitInputMethodMixin,

        DisabledPropertyMixin,
        LoadingPropertyMixin,
        NamePropertyMixin,
        ParentFormGroupPropertyMixin,
        ReadonlyPropertyMixin,
        SizePropertyMixin,
        TabIndexPropertyMixin
    ],
    created() {
        this.classesProvider.add(() => ({
            '-inline': this.inline,
            '-error': this.inputSchema && this.inputSchema.invalid
        }));
    },
    watch: {
        'inputSchema.invalid' (invalid) {
            if (this.validate && invalid) {
                this.errors = Object.keys(this.inputSchema.errors)
                    .filter((key) => key !== 'length')
                    .map((key) => this.inputSchema.errors[key]);
            }
        }
    },
    methods: {
        clearErrors () {
            this.errors = [];
        }
    }
};
