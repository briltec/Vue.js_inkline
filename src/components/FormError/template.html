<div class="form-group" :class="classes" role="inputgroup">
    <slot></slot>

    <transition v-if="validate && inputSchema" name="fade-in-transition" @after-leave="clearErrors()">
        <slot name="error" v-if="inputSchema.errors.length > 0" >
            <ul class="form-group-error">
                <li v-for="error in errors">{{ error }}</li>
            </ul>
        </slot>
    </transition>
</div>