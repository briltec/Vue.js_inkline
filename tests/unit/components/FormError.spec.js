import { shallowMount } from '@vue/test-utils';
import FormError from '@inkline/inkline/components/FormError';

describe('Components', () => {
    describe('FormError', () => {
        let wrapper;

        beforeEach(() => {
            wrapper = shallowMount(FormError);
        });

        it('should be named correctly', () => {
            expect(FormError.name).toEqual('IFormError');
        });

        it('should render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('props', () => {
            describe('schema', () => {
                it('should be defined', () => {
                    expect(wrapper.vm).toHaveProperty('schema');
                });
            });
        });

        describe('data()', () => {
            describe('errors', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.errors).toBeDefined();
                    expect(wrapper.vm.errors).toEqual([]);
                });
            });
        });

        describe('watch', () => {
            describe('schema.invalid()', () => {
                it('should add schema error messages to errors array', () => {
                    wrapper.setProps({
                        schema: {
                            invalid: false
                        }
                    });

                    expect(wrapper.vm.errors).toEqual([]);

                    wrapper.setProps({
                        schema: {
                            errors: {
                                errorA: 'a',
                                errorB: 'b'
                            },
                            invalid: true
                        }
                    });


                    expect(wrapper.vm.errors).toEqual(['a', 'b']);
                });
            });
        });

        describe('methods', () => {
            describe('clearErrors()', () => {
                it('should clear error messages', () => {
                    wrapper.setData({ errors: ['a', 'b', 'c'] });
                    wrapper.clearErrors();

                    expect(wrapper.vm.errors).toEqual([]);
                });
            });
        });
    });
});
