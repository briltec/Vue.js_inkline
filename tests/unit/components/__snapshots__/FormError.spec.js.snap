// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components FormError should render correctly 1`] = `undefined`;
